internal/admission/eventsource/handler_test.go

```diff
@@ -0,0 +1,155 @@
+package eventsource_test
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"testing"
+
+	"github.com/kanopy-platform/argoslower/internal/admission/eventsource"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	admissionv1 "k8s.io/api/admission/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	common "github.com/argoproj/argo-events/pkg/apis/common"
+	esv1alpha1 "github.com/argoproj/argo-events/pkg/apis/eventsource/v1alpha1"
+)
+
+type FakeMeshChecker struct {
+	Mesh bool
+	Err  error
+}
+
+func (m *FakeMeshChecker) OnMesh(ns string) (bool, error) {
+	return m.Mesh, m.Err
+}
+
+func TestEventSourceHandler(t *testing.T) {
+
+	t.Parallel()
+	fmc := &FakeMeshChecker{
+		Mesh: true,
+	}
+
+	handler := eventsource.NewHandler(fmc)
+	scheme := runtime.NewScheme()
+	utilruntime.Must(esv1alpha1.AddToScheme(scheme))
+	decoder, err := admission.NewDecoder(scheme)
+	assert.NoError(t, err)
+	err = handler.InjectDecoder(decoder)
+	assert.NoError(t, err)
+
+	f := false
+	for _, test := range []struct {
+		name     string
+		es       esv1alpha1.EventSource
+		key      string
+		err      bool
+		nsOnMesh *bool
+		nsErr    error
+	}{
+		{
+			name: "No Annotation",
+		},
+		{
+			name: "Opted In, labels",
+			es: esv1alpha1.EventSource{
+				ObjectMeta: v1.ObjectMeta{
+					Annotations: map[string]string{
+						eventsource.DefaultAnnotationKey: "true",
+					},
+				},
+				Spec: esv1alpha1.EventSourceSpec{
+					Template: &esv1alpha1.Template{
+						Metadata: &common.Metadata{
+							Labels: map[string]string{
+								"sidecar.istio.io/inject": "false",
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "Opted In, no labels",
+			es: esv1alpha1.EventSource{
+				ObjectMeta: v1.ObjectMeta{
+					Annotations: map[string]string{
+						eventsource.DefaultAnnotationKey: "false",
+					},
+				},
+			},
+		},
+		{
+			name: "Namespace Opted Out",
+			es: esv1alpha1.EventSource{
+				ObjectMeta: v1.ObjectMeta{
+					Annotations: map[string]string{
+						eventsource.DefaultAnnotationKey: "true",
+					},
+				},
+			},
+			nsOnMesh: &f,
+		},
+		{
+			name: "Namespace Error",
+			es: esv1alpha1.EventSource{
+				ObjectMeta: v1.ObjectMeta{
+					Annotations: map[string]string{
+						eventsource.DefaultAnnotationKey: "true",
+					},
+				},
+			},
+			nsErr: errors.New("test error"),
+		},
+	} {
+		if test.key == "" {
+			test.key = eventsource.DefaultAnnotationKey
+		}
+
+		if test.nsOnMesh != nil {
+			fmc.Mesh = false
+		}
```

Comment:
https://github.com/kanopy-platform/argoslower/pull/12#discussion_r1438293241
The sequence for fmc.Mesh is weird. It gets set here for testing, then reset on L130-L131.

Would it make more sense to have fmc in the test struct. If nil here, use the default. Then handler.meshChecker = test.fmc.

Reply:
@yuzhouliu9 I would prefer to not have to do all the handler instantiation over again for each loop so I am manipulating the FakeMeshChecker. 

I am in a different package than the handler `package eventsource_test` so I cannot call private functions.